### PR TITLE
fix: `on_fit_epoch_end` in `final_eval` overwriting last epoch metrics

### LINKED ISSUE #18735: Callback on_fit_epoch_end called twice at the end of model.train
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

Hi!

I noticed that the `on_fit_epoch_end` callback event gets triggered also at the end of a training run in the final_eval: https://github.com/ultralytics/ultralytics/blob/d9292fb7f10a2ee0265af28fad5d56b8ab79eece/ultralytics/engine/trainer.py#L689

I am wondering if that's intended. This leads to problems when using the call back for example log mlflow metrics, because the eval metric is being logged in the same graph as the metrics during training, with a duplicate epoch:
Image

Note I am not sure how this looks like if you're using the ultralytics mlflow integration, we decided not to use it because we failed to log everything in one run, ultralytics would always create a seperate run even though we run training within an mlflow run context already. So I think there might be another problem.

Could this be changed to have a separate callback event for this?
Happy to contribute if you agree.

### Environment

```
Ultralytics 8.3.62 🚀 Python-3.12.3 torch-2.4.0+cu124 CUDA:0 (Tesla T4, 14931MiB)
Setup complete ✅ (4 CPUs, 10.8 GB RAM, 36.8/146.6 GB disk)

OS                  Linux-5.15.0-1075-aws-x86_64-with-glibc2.39
Environment         Linux
Python              3.12.3
Install             pip
RAM                 10.83 GB
Disk                36.8/146.6 GB
CPU                 Intel Xeon Platinum 8259CL 2.50GHz
CPU count           4
GPU                 Tesla T4, 14931MiB
GPU count           1
CUDA                12.4

numpy               ✅ 1.26.4>=1.23.0
numpy               ✅ 1.26.4<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.8.4>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.2>=2.23.0
scipy               ✅ 1.13.1>=1.4.1
torch               ✅ 2.4.0+cu124>=1.8.0
torch               ✅ 2.4.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.19.0+cu124>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 5.9.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 1.5.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0
```

### Minimal Reproducible Example
Note you have to have a local mlflow tracking server running.

```
import mlflow
from ultralytics import YOLO, settings

def main():
    remote_server_uri = "http://localhost:8080"  # set to your server URI
    mlflow.set_tracking_uri(remote_server_uri)
    mlflow.set_experiment("/my-experiment")

    mlflow.pytorch.autolog(disable=False)
    settings.update({"mlflow": False})

    def sanitize_dict(x):
        """Sanitize dictionary keys by removing parentheses and converting values to floats."""
        return {"cb_" + k.replace("(", "").replace(")", ""): float(v) for k, v in x.items()}
    
    def on_fit_epoch_end(trainer):
        print(f"{trainer.epoch}, {trainer.metrics}")
        if mlflow:
            mlflow.log_metrics(metrics=sanitize_dict(trainer.metrics), step=trainer.epoch)

    with mlflow.start_run():
        # Create a new YOLO model from scratch
        model = YOLO("yolo11n.pt")

        # Load a pretrained YOLO model (recommended for training)
        model.add_callback("on_fit_epoch_end", on_fit_epoch_end)

        # Train the model using the 'coco8.yaml' dataset for 3 epochs
        results = model.train(data="coco8.yaml", epochs=3)

if __name__ == "__main__":
    main()

```
### Additional

_No response_

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### LINKED ISSUE #20864: TensorBoard logging error of last epoch.
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

_No response_

### Bug

During the detailed validation that happens after the training, TensorBoard log of the last epoch mAP50-95 and last epoch mAP50 get overwritten by the best checkpoint value (instead of last checkpoint value).

I have reported this issue to Giovanni (giovanni.dalzilio@ultralytics.com).
It has already beed confirmed as a bug.

Here is a portion of the original email regarding a 100 epochs training:

> I have noticed an bug in TensorBoard logging of the last epoch (epoch number 100) in ultralytics pipeline (such issue did not occur in yolov5 pipeline).
> In all training histories in the console output and in the csv file, I can see that the mAP50-95 continues dropping down in the last epoch.
> Just before the final validation happens, it is displayed in TensorBoard correctly for a moment....
> However when the summary validation happens in the end of training, the result of epoch 100 is replaced in TensorBoard to value of best checkpoint and value of epoch 100 disappears.
> It seems like there is a bug in implementation.
> In yolov5 repository there was no such issue. Also epochs were indexed from 0 to 99 instead of 1 to 100.
> My hypothesis is that it is:
> - either completely unintentional overwriting (in such scenario it should not be logged in training),
> - or someone wanted to visualise the top value in the end after the last epoch as a feature for easier comparisons of top scores of many trainings - here it replaces the last epoch instead of being placed after it.

### Environment

Ultralytics 8.3.145  Python-3.12.3 torch-2.4.1+cu121 CUDA:0 (NVIDIA GeForce RTX 4090, 24217MiB)
Setup complete ✅ (32 CPUs, 124.9 GB RAM, 1229.9/1829.7 GB disk)

OS                  Linux-6.11.0-25-generic-x86_64-with-glibc2.39
Environment         Linux
Python              3.12.3
Install             git
Path                /home/karol-badowski/Development/ultralytics/ultralytics
RAM                 124.90 GB
Disk                1229.9/1829.7 GB
CPU                 AMD Ryzen 9 7950X3D 16-Core Processor
CPU count           32
GPU                 NVIDIA GeForce RTX 4090, 24217MiB
GPU count           1
CUDA                12.1

numpy               ✅ 1.26.4>=1.23.0
matplotlib          ✅ 3.9.1>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.0>=1.4.1
torch               ✅ 2.4.1>=1.8.0
torch               ✅ 2.4.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.19.0>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.0>=2.0.0

### Minimal Reproducible Example

Just run a default training.

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by deependujha**
### Description
`final_eval()` calls `run_callbacks("on_fit_epoch_end")` with `self.epoch` unchanged. Since the training loop already fired the same callback at this epoch value, any logger using `trainer.epoch` as a step index (TensorBoard, MLflow, etc.) overwrites the last epoch's actual metrics with best-checkpoint values instead of appending after them.

### Root Cause
- Training loop fires `on_fit_epoch_end` at `self.epoch = N` (last epoch).
- `final_eval()` then fires `on_fit_epoch_end` again at `self.epoch = N` → collision.

### Fix
Increment `self.epoch` by 1 before the callback in `final_eval`, restore after:

```python
self.epoch += 1  # avoid overwriting last epoch's metrics in loggers
self.run_callbacks("on_fit_epoch_end")
self.epoch -= 1  # restore
```

- Best metrics are now logged at step `epochs+1`, a distinct trailing point in TensorBoard/MLflow.
- The existing behavior of surfacing best metrics via `on_fit_epoch_end` (noted in #18804) is fully preserved, just at the correct step.

### Behavior on resumed training
- If training resumes from a checkpoint at epoch `N`, the saved `on_fit_epoch_end` at step `N+1` (best metrics from previous run) gets overwritten by the first real epoch of the resumed run.
- This is actually correct, the resumed run's actual epoch metrics should take precedence over the stale best-metrics marker from the previous run. 
- At the end of the resumed training (epoch `M`), `final_eval` writes best metrics at `M+1` as expected.

### Related Issues
- Fixes #20864
- Fixes #18735
- Ref #18804

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Fixes a training callback logging bug where `final_eval()` could overwrite the last epoch's metrics by temporarily shifting the epoch index before running `on_fit_epoch_end`. 🛠️📈

### 📊 Key Changes
- Updates `ultralytics/engine/trainer.py` in `final_eval()`.
- Increments `self.epoch` before calling `self.run_callbacks("on_fit_epoch_end")`.
- Decrements `self.epoch` immediately afterward to restore the original state.
- Preserves the existing final validation flow, including removing `fitness` from final metrics.

### 🎯 Purpose & Impact
- Prevents best/final evaluation metrics from being logged on top of the last training epoch metrics.
- Ensures callback consumers receive final-eval metrics at a distinct step after training completes.
- Improves metric tracking consistency for training logs, visualizations, and downstream integrations. ✅

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @deependujha, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help streamline review, and an engineer will assist you shortly. Please review the checklist below for a smooth integration 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The fix addresses the metric step-collision bug cleanly, but the temporary epoch shift should be exception-safe. Wrapping the callback in `try/finally` avoids leaving `self.epoch` incremented when a callback fails.

💬 Posted 1 inline comment

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The change looks clean and directly addresses the reported metric-step collision by ensuring the final `on_fit_epoch_end` callback runs at a distinct step while safely restoring `self.epoch` in a `finally` block. I did not find any clear bug or regression risk in the diff shown.

**Comment by Laughing-q** [comment]
@deependujha yeah I encountered the same with wandb logging as well, now the latest wandb package does not support overriding anymore so the best pt results are not logged correctly with wandb. This fix might be the best/simplest way to do it now I think. THank you!

**Comment by Laughing-q** [comment]
@deependujha merging! thank you for your contribution!
